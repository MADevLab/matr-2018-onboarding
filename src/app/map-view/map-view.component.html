<div id="container" style="min-height: 600px; width: 100%;"></div>
