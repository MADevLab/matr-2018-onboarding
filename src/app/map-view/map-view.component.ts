import { Component, Input, OnChanges } from '@angular/core';
import { DataPoints, DataPoint } from '../data-set.model';
import { Regions } from '../regions.model';

declare var Highcharts: any;

@Component({
  selector: 'app-map-view',
  templateUrl: './map-view.component.html'
})
export class MapViewComponent implements OnChanges {
  @Input() public dataPoints: DataPoints;

  @Input() public title: string = null;

  @Input() public concept: string = null;

  @Input() public selectedRegions: Regions[];

  /**
   * Highmap instance
   *
   * @type {*}
   * @memberof AppComponent
   */
  public map: any;

  constructor() {}

  ngOnChanges() {
    if (this.map) {
      this.map.destroy();
    }

    throw new Error('Missing Logic');

    const regions = this.getStates(this.selectedRegions);

    if (regions && regions.length) {
      // uncomment once data logic is implemented
      //data = data.filter(({ code }) => regions.includes(code));
    }

    // uncomment once data logic is implemented
    //this.renderHighmap(data);
  }

  public getStates(regions: Regions[]): string[] {
    throw new Error('Missing Logic');
  }

  public renderHighmap(data: any) {
    const el = document.getElementById('container');

    if (!el) {
      throw new Error('A div with ID of container is required');
    }

    if (typeof Highcharts !== 'undefined') {

      // Instantiate the map
      this.map = Highcharts.mapChart(el, {
        chart: {
          map: 'countries/us/us-all',
          borderWidth: 1
        },

        title: {
          text: this.title
        },

        legend: {
          layout: 'horizontal',
          borderWidth: 0,
          backgroundColor: 'rgba(255,255,255,0.85)',
          floating: true,
          verticalAlign: 'top',
          y: 25
        },

        mapNavigation: {
          enabled: false
        },

        colorAxis: {
          min: null,
          type: 'linear',
          minColor: '#FF5722',
          maxColor: '#00bcd4'
        },

        series: [
          {
            animation: false,
            data: data,
            joinBy: ['postal-code', 'code'],
            dataLabels: {
              enabled: true,
              color: '#FFFFFF',
              format: '{point.code}'
            },
            name: this.concept,
            tooltip: {
              pointFormat: '{point.code}: {point.value}'
            }
          }
        ]
      });
    }
  }

  private formatMapData = (dataPoint: DataPoint) => {
    return Object.assign({}, dataPoint, { code: dataPoint.code.toUpperCase() });
  }
}
