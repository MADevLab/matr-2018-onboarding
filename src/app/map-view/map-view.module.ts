import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MapViewComponent } from './map-view.component';

@NgModule({
  imports: [CommonModule],
  declarations: [MapViewComponent],
  exports: [MapViewComponent]
})
export class MapViewModule {}
