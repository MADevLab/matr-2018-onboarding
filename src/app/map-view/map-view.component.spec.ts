import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { MapViewComponent } from './map-view.component';

describe('MapViewComponent', () => {
  let component: MapViewComponent;
  let fixture: ComponentFixture<MapViewComponent>;
  let renderSpy: jest.Mock<any>;

  beforeEach(async () => {
    TestBed.configureTestingModule({
      declarations: [ MapViewComponent ]
    });

    await TestBed.compileComponents();

    fixture = TestBed.createComponent(MapViewComponent);
    component = fixture.componentInstance;
    renderSpy = jest.spyOn(component, 'renderHighmap').mockImplementation();
    component.dataPoints = _data;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('produces an array of states based on the regions', () => {
    const result = component.getStates(_regions);

    expect(result).toMatchSnapshot();
  });

  describe('on change', () => {
    it('formats the data and calls render', () => {
      component.ngOnChanges();

      expect(renderSpy).toHaveBeenCalled();
      expect(renderSpy).toMatchSnapshot();
    });

    it('filters the data', () => {
      const spy = jest.spyOn(component, 'getStates');

      component.selectedRegions = _regions;
      component.ngOnChanges();

      expect(spy).toHaveBeenCalled();
      expect(spy).toMatchSnapshot();
    });

    it('filters the data based on the regions', () => {
      component.selectedRegions = _regions;
      component.ngOnChanges();

      expect(renderSpy).toHaveBeenCalled();
      expect(renderSpy).toMatchSnapshot();
    });

    it('destroys the map, if it exists', () => {
      const spy = jest.fn();

      component.map = {
        destroy: spy
      };
      component.ngOnChanges();

      expect(spy).toHaveBeenCalled();
    });
  });
});

const _regions = [
  {
    id: 'NE',
    title: 'Northeast',
    regions: ['CT', 'DC', 'DE']
  },
  {
    id: 'MW',
    title: 'Midwest',
    regions: ['WI']
  },
  {
    id: 'SO',
    title: 'South',
    regions: ['AL', 'AR']
  },
  {
    id: 'WE',
    title: 'West',
    regions: ['WA', 'WY']
  }
];

const _data = {
  period: 661,
  date: '2015Q1',
  mapData: [
    { value: 18.1, code: 'ak' },
    { value: 84.58, code: 'al' },
    { value: 44.47, code: 'ar' },
    { value: 112.52, code: 'az' },
    { value: 375.38, code: 'ca' },
    { value: 101.97, code: 'co' },
    { value: 35.69, code: 'ct' },
    { value: 6.36, code: 'dc' },
    { value: 9.6, code: 'de' },
    { value: 435.75, code: 'fl' },
    { value: 142.13, code: 'ga' },
    { value: 16.38, code: 'hi' },
    { value: 53.2, code: 'ia' },
    { value: 36.03, code: 'id' },
    { value: 185.43, code: 'il' },
    { value: 118.85, code: 'in' },
    { value: 38.62, code: 'ks' },
    { value: 62.43, code: 'ky' },
    { value: 55.09, code: 'la' },
    { value: 71.14, code: 'ma' },
    { value: 70.91, code: 'md' },
    { value: 16.34, code: 'me' },
    { value: 156.56, code: 'mi' },
    { value: 86, code: 'mn' },
    { value: 97.64, code: 'mo' },
    { value: 39.08, code: 'ms' },
    { value: 16.16, code: 'mt' },
    { value: 134.4, code: 'nc' },
    { value: 12.5, 'code': 'nd' },
    { value: 33.42, code: 'ne' },
    { value: 18.34, code: 'nh' },
    { value: 89.13, code: 'nj' },
    { value: 24.98, code: 'nm' },
    { value: 65.87, code: 'nv' },
    { value: 143.63, code: 'ny' },
    { value: 223.17, code: 'oh' },
    { value: 62.36, code: 'ok' },
    { value: 62.43, code: 'or' },
    { value: 141.63, code: 'pa' },
    { value: 11.17, code: 'ri' },
    { value: 77.15, code: 'sc' },
    { value: 5.16, code: 'sd' },
    { value: 110.86, code: 'tn' },
    { value: 399.95, code: 'tx' },
    { value: 54.35, code: 'ut' },
    { value: 106.38, code: 'va' },
    { value: 9.6, code: 'vt' },
    { value: 103.66, code: 'wa' },
    { value: 86.11, code: 'wi' },
    { value: 25.16, code: 'wv' },
    { value: 8.84, code: 'wy' }
  ]
};
