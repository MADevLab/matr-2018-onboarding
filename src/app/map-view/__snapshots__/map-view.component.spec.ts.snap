// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapViewComponent on change filters the data 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "id": "NE",
          "regions": Array [
            "CT",
            "DC",
            "DE",
          ],
          "title": "Northeast",
        },
        Object {
          "id": "MW",
          "regions": Array [
            "WI",
          ],
          "title": "Midwest",
        },
        Object {
          "id": "SO",
          "regions": Array [
            "AL",
            "AR",
          ],
          "title": "South",
        },
        Object {
          "id": "WE",
          "regions": Array [
            "WA",
            "WY",
          ],
          "title": "West",
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Array [
        "CT",
        "DC",
        "DE",
        "WI",
        "AL",
        "AR",
        "WA",
        "WY",
      ],
    },
  ],
}
`;

exports[`MapViewComponent on change filters the data based on the regions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "code": "AL",
          "value": 84.58,
        },
        Object {
          "code": "AR",
          "value": 44.47,
        },
        Object {
          "code": "CT",
          "value": 35.69,
        },
        Object {
          "code": "DC",
          "value": 6.36,
        },
        Object {
          "code": "DE",
          "value": 9.6,
        },
        Object {
          "code": "WA",
          "value": 103.66,
        },
        Object {
          "code": "WI",
          "value": 86.11,
        },
        Object {
          "code": "WY",
          "value": 8.84,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`MapViewComponent on change formats the data and calls render 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        Object {
          "code": "AK",
          "value": 18.1,
        },
        Object {
          "code": "AL",
          "value": 84.58,
        },
        Object {
          "code": "AR",
          "value": 44.47,
        },
        Object {
          "code": "AZ",
          "value": 112.52,
        },
        Object {
          "code": "CA",
          "value": 375.38,
        },
        Object {
          "code": "CO",
          "value": 101.97,
        },
        Object {
          "code": "CT",
          "value": 35.69,
        },
        Object {
          "code": "DC",
          "value": 6.36,
        },
        Object {
          "code": "DE",
          "value": 9.6,
        },
        Object {
          "code": "FL",
          "value": 435.75,
        },
        Object {
          "code": "GA",
          "value": 142.13,
        },
        Object {
          "code": "HI",
          "value": 16.38,
        },
        Object {
          "code": "IA",
          "value": 53.2,
        },
        Object {
          "code": "ID",
          "value": 36.03,
        },
        Object {
          "code": "IL",
          "value": 185.43,
        },
        Object {
          "code": "IN",
          "value": 118.85,
        },
        Object {
          "code": "KS",
          "value": 38.62,
        },
        Object {
          "code": "KY",
          "value": 62.43,
        },
        Object {
          "code": "LA",
          "value": 55.09,
        },
        Object {
          "code": "MA",
          "value": 71.14,
        },
        Object {
          "code": "MD",
          "value": 70.91,
        },
        Object {
          "code": "ME",
          "value": 16.34,
        },
        Object {
          "code": "MI",
          "value": 156.56,
        },
        Object {
          "code": "MN",
          "value": 86,
        },
        Object {
          "code": "MO",
          "value": 97.64,
        },
        Object {
          "code": "MS",
          "value": 39.08,
        },
        Object {
          "code": "MT",
          "value": 16.16,
        },
        Object {
          "code": "NC",
          "value": 134.4,
        },
        Object {
          "code": "ND",
          "value": 12.5,
        },
        Object {
          "code": "NE",
          "value": 33.42,
        },
        Object {
          "code": "NH",
          "value": 18.34,
        },
        Object {
          "code": "NJ",
          "value": 89.13,
        },
        Object {
          "code": "NM",
          "value": 24.98,
        },
        Object {
          "code": "NV",
          "value": 65.87,
        },
        Object {
          "code": "NY",
          "value": 143.63,
        },
        Object {
          "code": "OH",
          "value": 223.17,
        },
        Object {
          "code": "OK",
          "value": 62.36,
        },
        Object {
          "code": "OR",
          "value": 62.43,
        },
        Object {
          "code": "PA",
          "value": 141.63,
        },
        Object {
          "code": "RI",
          "value": 11.17,
        },
        Object {
          "code": "SC",
          "value": 77.15,
        },
        Object {
          "code": "SD",
          "value": 5.16,
        },
        Object {
          "code": "TN",
          "value": 110.86,
        },
        Object {
          "code": "TX",
          "value": 399.95,
        },
        Object {
          "code": "UT",
          "value": 54.35,
        },
        Object {
          "code": "VA",
          "value": 106.38,
        },
        Object {
          "code": "VT",
          "value": 9.6,
        },
        Object {
          "code": "WA",
          "value": 103.66,
        },
        Object {
          "code": "WI",
          "value": 86.11,
        },
        Object {
          "code": "WV",
          "value": 25.16,
        },
        Object {
          "code": "WY",
          "value": 8.84,
        },
      ],
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`MapViewComponent produces an array of states based on the regions 1`] = `
Array [
  "CT",
  "DC",
  "DE",
  "WI",
  "AL",
  "AR",
  "WA",
  "WY",
]
`;
